lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -1368,11 +1041,16 @@ def list_from_ruby(arr)
       ret
     end
 
-    def struct_from_ruby(hash)
+    # If reduce is true, nil will be returned for hash like
+    # {"httpRequest"=>{"referer"=>nil}} instead of layers of empty structs.
+    def struct_from_ruby(hash, reduce = true)
       ret = Google::Protobuf::Struct.new
       hash.each do |k, v|
-        ret.fields[k] ||= value_from_ruby(v)
+        ret.fields[k] ||= value_from_ruby(v) \
+          if !v.nil? && !value_from_ruby(v).nil?
```

Comment:
Basically, the reason we had layered empty structs in 'httpRequest' is because:
When user specify 'referer=nil', in a `google.protobuf.Struct` it's presented as `{"nullValue"=>"NULL_VALUE"}`. And in the previous implementation, `referer` is not being deleted from `httpRequest` even though it's technically nil:

```
  output.referer = input.delete('referer') unless
    input['referer'].nil?
```

As a result, when we check if `input` is empty, we did not delete `httpRequest` properly:

```
record.delete('httpRequest') if input.empty?
```